Fix wrong model after login (#213)

hassoncs opened the pull request:
Previously we were spreading the entire `prevCachedState.apiConfiguration` object when the keys changed, which would wipe out unsaved changes the user might have made to things like `prevCachedState.apiConfiguration.kiloModel`. Now we only set the keys that might be set via callback and don't trigger a save explicitly. The user will need to click 'save' in order to exit the SettingView like normal (same as changing any other setting).

![213-fix-wrong-model-after-login](https://github.com/user-attachments/assets/4b2381d3-22ca-414c-a12b-f65940ef8749)


- update state with specific tokens/keys to avoid overwriting unsaved changes
- simplify useEffect dependencies by using direct variables

**Dev thoughts**
This does make me think.. should "login" keys / tokens be a "setting"? seems like it should be in the Settings view UI, but possibly separate from the other settings state of the api. I feel like it's not intuitive to have to click a "save" button once you login to something for it to remain saved.

**Test plan**: Verified that logging in / out no longer overwrites other non-saved changes in the Settings. Now users will have to click save for any changes they make, including login state. (This should be ok since we have that confirmation dialog if they don't.)

- Fixes #213

hassoncs committed:
🐞 fix(settings): Fix Wrong model after login (#213)

Previously we were spreading the entire `prevCachedState.apiConfiguration` object when the keys changed, which would wipe out unsaved changes the user might have made to things like `prevCachedState.apiConfiguration.kiloModel`. Now we only set the keys that might be set via callback and don't trigger a save explicitly. The user will need to click 'save' in order to exit the SettingView like normal (same as changing any other setting).

- update state with specific tokens/keys to avoid overwriting unsaved changes
- simplify useEffect dependencies by using direct variables

**Dev thoughts**
This does make me think.. should "login" keys / tokens be a "setting"? seems like it should be in the Settings view UI, but possibly separate from the other settings state of the api. I feel like it's not intuitive to have to click a "save" button once you login to something for it to remain saved.

**Test plan**: Verified that logging in / out no longer overwrites other non-saved changes in the Settings. Now users will have to click save for any changes they make, including login state. (This should be ok since we have that confirmation dialog if they don't.)

**Closes** #123

drakonen reviewed (commented):
Looks good to me, unsure about the .changeset file if that should be included,

@kevinvandijk should this PR have a .changeset file?

kevinvandijk reviewed (changes requested):
This introduces a regression of a previous bug. Now when changing a setting and then clicking on 'Done', then clicking on 'Discard changes'. It will freeze. See screen shot 
<img width="336" alt="image" src="https://github.com/user-attachments/assets/c2aa2f91-a8bd-4280-8bff-676695c270de" />

hassoncs committed:
🐛 fix(settings): fix bug where "discard changes" would freeze the SettingsView

- remove setTimeout wrapping confirmDialogHandler call for immediate execution

I've fixed the bug in the settings view that was causing the "Discard unsaved changes" dialog to appear twice and freeze when changing the diffEnabled setting.

The issue was a race condition in the onConfirmDialogResult function. When using setTimeout to delay the callback execution, React would re-render the component before the callback ran. During this re-render, the effect comparing cachedState and extensionState would detect a difference again specifically with the diffEnabled setting, causing isChangeDetected to be set back to true and triggering an infinite loop.

The fix removes the setTimeout and calls the callback directly after setting the state, ensuring it runs before any effects can set isChangeDetected back to true.

Test plan:
- Logged in / out confirmed unsaved settings do not change
- update a setting, discard, confirm no freeze
- update setting, save, confirm saved

hassoncs commented:
good catch, @kevinvandijk! It took me a second to figure out, but it seems the bug you found was already on the main branch. I went back to `main` and I was able to reproduce it. It seemed to be caused by the `setTimeout` in the dialog code. I added a new commit to fix it with an explanation. I tested it pretty well, but it wouldn't hurt to confirm that I didn't miss anything!

kevinvandijk reviewed (approved):
Nice find @hassoncs! I was not aware it was in main as well. Seems to work well now here indeed!
